winch/codegen/src/isa/x64/masm.rs

```diff
@@ -206,6 +206,30 @@ impl Masm for MacroAssembler {
         context.stack.push(TypedReg::i32(size).into());
     }
 
+    fn memory_size(&mut self, heap_data: &crate::codegen::HeapData, context: &mut CodeGenContext) {
+        let size_reg = context.any_gpr(self);
+        let scratch = regs::scratch();
+        let vmctx = <Self::ABI as ABI>::vmctx_reg();
+
+        let base = if let Some(offset) = heap_data.import_from {
+            self.asm
+                .mov_mr(&self.address_at_vmctx(offset), scratch, self.ptr_size);
+            scratch
+        } else {
+            vmctx
+        };
+
+        let size_addr = Address::offset(base, heap_data.current_length_offset);
+        self.asm.mov_mr(&size_addr, size_reg, self.ptr_size);
+        // Prepare the stack to emit division and get the size in pages rather
+        // than in bytes.
+        context
+            .stack
+            .push(TypedReg::new(heap_data.ty, size_reg).into());
+        context.stack.push(Val::i64(WASM_PAGE_SIZE as i64));
+        self.div(context, DivKind::Unsigned, heap_data.ty.into());
```

Comment:
Given that the page size is a power-of-two could this be hardcoded as a right-shift instead of a division?

Reply:
Indeed, fixed! Thanks for the suggestion, using a shift should also help with register pressure (at least for x64).